crates/accelerate/src/unitary_synthesis.rs

```diff
@@ -0,0 +1,1094 @@
+        embodiments_dict.set_item(strength, embodiment.into_py(py))?;
+    }
+
+    // Iterate over 2q fidelities ans select decomposers
+    if not_empty {
+        let xx_decomposer: Bound<'_, PyAny> =
+            PyModule::import_bound(py, "qiskit.synthesis.two_qubit.xx_decompose")?
+                .getattr("XXDecomposer")?;
+
+        for basis_1q in &available_1q_basis {
+            let pi2_decomposer = if let Some(pi_2_basis) = pi2_basis {
+                if pi_2_basis == "cx" && *basis_1q == "ZSX" {
+                    let fidelity = match approximation_degree {
+                        Some(approx_degree) => approx_degree,
+                        None => match &target["cx"][Some(&qubits)] {
+                            Some(props) => 1.0 - props.error.unwrap_or(0.0),
+                            None => 1.0,
+                        },
+                    };
+                    Some(TwoQubitBasisDecomposer::new_inner(
+                        pi_2_basis.to_string(),
+                        StandardGate::CXGate.matrix(&[]).unwrap().view(),
+                        fidelity,
+                        basis_1q,
+                        Some(true),
+                    )?)
+                } else {
+                    None
+                }
+            } else {
+                None
+            };
+
+            let decomposer = xx_decomposer.call1((
+                &basis_2q_fidelity_dict,
+                PyString::new_bound(py, basis_1q),
+                &embodiments_dict,
+                pi2_decomposer,
+            ))?;
+            let decomposer_gate = decomposer.getattr("gate")?;
+
+            // .getattr("name")?
+            // .extract::<String>()?;
+
+            decomposers.push(DecomposerElement {
+                decomposer: DecomposerType::XXDecomposer(decomposer.into()),
+                decomp_gate: decomposer_gate.getattr("name")?.extract::<String>()?,
+                decomp_gate_params: Some(
+                    decomposer_gate
+                        .getattr("params")?
+                        .extract::<SmallVec<[Param; 3]>>()?,
+                ),
+            });
+        }
+    }
+    Ok(Some(decomposers))
+}
+
+fn preferred_direction(
+    decomposer: &DecomposerElement,
+    wire_map: &IndexMap<Qubit, PhysicalQubit>,
+    natural_direction: Option<bool>,
+    neighbor_table: &Option<&NeighborTable>,
+    distance_matrix: &Option<PyReadonlyArray2<f64>>,
+    target: &Target,
+) -> PyResult<Option<bool>> {
+    // Returns:
+    // * true if gate qubits are in the hardware-native direction
+    // * false if gate qubits must be flipped to match hardware-native direction
+
+    let qubits: SmallVec<[PhysicalQubit; 2]> = wire_map.values().copied().collect();
+    let reverse_qubits: SmallVec<[PhysicalQubit; 2]> = qubits.iter().rev().copied().collect();
+
+    let compute_cost =
+        |lengths: bool, q_tuple: &SmallVec<[PhysicalQubit; 2]>, in_cost: f64| -> PyResult<f64> {
```

Comment:
You don't actually need a smallvec here it can be any slice. Not that it'll make a big difference, but if you make my above suggestion about the typing this can be anything that can be a slice:

```suggestion
        |lengths: bool, q_tuple: &[PhysicalQubit; 2], in_cost: f64| -> PyResult<f64> {
```